test/enhancedStepFunctions.js

```diff
@@ -851,144 +852,143 @@ contract('ConversionRates2', function(accounts) {
-        stepY = [-20, -10, -5];
-        bps = getExtraBpsForQuantity(-75, stepX, stepY);
-        // (-75 - (-50)) * (-5) + (-50 - 0) * (-5)
-        assert.equal(-5, bps, "bad bps");
+        stepY = [-20, -10, -5, -2];
+        bps = getExtraBpsForQuantity(-75, 10, stepX, stepY);
+        // 25 * (-5) + 60 * (-2)
+        assert.equal(divSolidity(-245, 85), bps, "bad bps");
         await convRatesInst.setImbalanceStepFunction(token, stepX, stepY, stepX, stepY, {from:operator});
-        contractBps = await convRatesInst.mockExecuteStepFunction(token, -75);
+        contractBps = await convRatesInst.mockExecuteStepFunction(token, -75, 10);
         assert.equal(contractBps.valueOf(), bps, "bad bps");
 
         // Case3: qty negative, all steps negative except last one is 0
         stepX = [-200, -100, -50, 0];
-        stepY = [-20, -10, -5, 1];
-        // (-300 - (-200)) * (-20) + (-200 - (-100)) * (-10) + (-100 - (-50)) * (-5) + (-50 - 0) * 1 = 3200
-        bps = getExtraBpsForQuantity(-300, stepX, stepY);
-        assert.equal(divSolidity(3200, -300), bps, "bad bps");
+        stepY = [-20, -10, -5, 1, 5];
+        // 100 * (-20) + 100 * (-10) + 50 * (-5) + 20 * 1 = -3230
+        bps = getExtraBpsForQuantity(-300, -25, stepX, stepY);
+        assert.equal(divSolidity(-3230, 275), bps, "bad bps 1");
         await convRatesInst.setImbalanceStepFunction(token, stepX, stepY, stepX, stepY, {from:operator});
-        contractBps = await convRatesInst.mockExecuteStepFunction(token, -300);
-        assert.equal(contractBps.valueOf(), bps, "bad bps");
+        contractBps = await convRatesInst.mockExecuteStepFunction(token, -300, -25);
+        assert.equal(contractBps.valueOf(), bps, "bad bps 2");
 
         // Case4: qty negative, all steps negative except last one is positive, qty < all steps
         stepX = [-200, -100, -50, 10];
-        stepY = [-20, -10, -5, 2];
-        // (-300 - (-200)) * (-20) + (-200 - (-100)) * (-10) + (-100 - (-50)) * (-5) + (-50 - 0) * 2 = 3150
-        bps = getExtraBpsForQuantity(-300, stepX, stepY);
-        assert.equal(divSolidity(3150, -300), bps, "bad bps");
+        stepY = [-20, -10, -5, 2, 5];
+        // 100 * (-20) + 100 * (-10) + 50 * (-5) + 60 * 2 + 5 * 5 = -3105
+        bps = getExtraBpsForQuantity(-300, 15, stepX, stepY);
+        assert.equal(divSolidity(-3105, 315), bps, "bad bps 1");
         await convRatesInst.setImbalanceStepFunction(token, stepX, stepY, stepX, stepY, {from:operator});
-        contractBps = await convRatesInst.mockExecuteStepFunction(token, -300);
-        assert.equal(contractBps.valueOf(), bps, "bad bps");
+        contractBps = await convRatesInst.mockExecuteStepFunction(token, -300, 15);
+        assert.equal(contractBps.valueOf(), bps, "bad bps 2");
 
         // Case5: qty negative, all steps negative except last one is positive, first step < qty < last step
-        // (-150 - (-100)) * (-10) + (-100 - (-50)) * (-5) + (-50 - 0) * 2 = 650
```

Comment:
using previous steps?
qty > last step

Reply:
Updated all comments for each test case here.